test/ConcurrentCacheTest.cpp

```diff
@@ -0,0 +1,282 @@
+  // stored under key 3.
+  auto res = a.computeOnce(3, waiting_function("3"s, 100));
+  t.stop();
+  ASSERT_EQ("3"s, *res._resultPointer);
+  ASSERT_FALSE(res._wasCached);
+  ASSERT_GE(t.msecs(), 100);
+  ASSERT_EQ(0ul, a.numPinnedElements());
+  ASSERT_EQ(1ul, a.numCachedElements());
+  // No other results currently being computed
+  ASSERT_TRUE(a.getStorage().wlock()->_inProgress.empty());
+
+  t.reset();
+  t.start();
+  // takes 0 msecs to compute, as the request is served from the cache.
+  // request a pin, the result should be read from the cache and upgraded
+  // to a pinned result.
+  auto res2 = a.computeOncePinned(3, waiting_function("3"s, 100));
+  t.stop();
+  // computing result again: still yields "3", was cached and takes 0
+  // milliseconds (result is read from cache)
+  ASSERT_EQ("3"s, *res2._resultPointer);
+  ASSERT_TRUE(res2._wasCached);
+  ASSERT_EQ(res._resultPointer, res2._resultPointer);
+  ASSERT_LE(t.msecs(), 5);
+  ASSERT_EQ(1ul, a.numPinnedElements());
+  ASSERT_EQ(0ul, a.numCachedElements());
+  ASSERT_TRUE(a.getStorage().wlock()->_inProgress.empty());
+}
+
+TEST(ConcurrentCache, concurrentComputation) {
+  auto a = SimpleAdapter(3ul);
+  std::atomic<bool> flag = false;
+  auto compute = [&a, &flag]() {
+    return a.computeOnce(3, waiting_function("3"s, 100, &flag));
+  };
+  auto resultFuture = std::async(std::launch::async, compute);
+  // the background thread is now computing for 100 milliseconds, wait for
+  // some time s.t. the computation has safely started
+  // note: This test might fail on a single-threaded system.
+  while (!flag) {
+  }
+  // now the background computation should be ongoing and registered as
+  // "in progress"
+  ASSERT_EQ(0ul, a.numCachedElements());
+  ASSERT_EQ(1ul, a.getStorage().wlock()->_inProgress.size());
+  ASSERT_TRUE(a.getStorage().wlock()->_inProgress.contains(3));
+
+  // this call waits for the background task to compute, and then fetches the
+  // result. After this call completes, nothing is in progress and the result
+  // is cached.
+  auto result = compute();
+  ASSERT_EQ(1ul, a.numCachedElements());
+  ASSERT_TRUE(a.getStorage().wlock()->_inProgress.empty());
+  ASSERT_EQ("3"s, *result._resultPointer);
+  ASSERT_FALSE(result._wasCached);
+  auto result2 = resultFuture.get();
+  ASSERT_EQ(result._resultPointer, result2._resultPointer);
+  ASSERT_FALSE(result2._wasCached);
+}
+
+TEST(ConcurrentCache, concurrentPinnedComputation) {
+  auto a = SimpleAdapter(3ul);
+  std::atomic<bool> flag = false;
+  auto compute = [&a, &flag]() {
+    return a.computeOncePinned(3, waiting_function("3"s, 100, &flag));
+  };
+  auto resultFuture = std::async(std::launch::async, compute);
+  // the background thread is now computing for 100 milliseconds, wait for
+  // some time s.t. the computation has safely started
+  // note: This test might fail on a single-threaded system.
+  // now the background computation should be ongoing and registered as
+  // "in progress"
```

Comment:
Is there any way to test whether the system we are running on supports threads?

Reply:
I have found a wait which should work in a single-threaded scenario.